### DISCUSSION TITLE
Why is derive `Insertable` not allowed on empty structs?

### DISCUSSION THREAD

**Discussion by hasezoey** [diesel-rs/diesel#3684] [Q&A] [answered]
as the title says, why is the derive `Insertable` not allowed on empty structs?

for example i have the following sql schema:
```sql
CREATE TABLE franchise(
  _id INTEGER NOT NULL PRIMARY KEY
);
CREATE TABLE franchise_names(
  _id INTEGER NOT NULL PRIMARY KEY,
  for INTEGER NOT NULL REFERENCES franchise(_id),
  name TEXT NOT NULL
);
```

and would like to have a insertable struct for `franchise`, which automatically generates the `_id` always and not have a workaround like `pub _id: Option<i32>` if `_id` is never meant to be manually set

for example i would expect `Insertable` to generate something along the lines of

```rust
#[derive(Debug, Clone, Queryable /* Insertable */)]
#[diesel(table_name=franchise)]
pub struct CreateFranchise {
	// pub _id: Option<i32>,
}

impl Insertable<franchise::table> for CreateFranchise {
	type Values =
		<(std::option::Option<diesel::dsl::Eq<franchise::_id, i32>>,) as Insertable<franchise::table>>::Values;
	fn values(
		self,
	) -> <(std::option::Option<diesel::dsl::Eq<franchise::_id, i32>>,) as Insertable<franchise::table>>::Values {
		return (None::<i32>.map(|x| franchise::_id.eq(x)),).values();
	}
}
```

diesel 2.1.0

**Comment by weiznich** [comment, marked as answer]
It's not really possible to generate this impl because a `derive` macro can only work with the tokens the derive is applied to. That means in that case we just don't know anything about the fields of the table there, at least other than the fields listed in your struct. Therefore it's probably impossible to generate the desired code. 

You likely want to use [this](https://docs.diesel.rs/2.1.x/diesel/query_builder/struct.IncompleteInsertStatement.html#method.default_values) method instead for inserting just the default values into a table.

**Reply by hasezoey** [reply]
yes, i ended up using `.default_values()`, but made this because it seemed weird to me (because initially i also wanted to have a similar structure across all tables and have it already available if the table should change)

if i understand your answer correctly, you mean that in a derive-macro you have no way of knowing something like field `_id`(from the example) exists, because you only have access to the current struct (in this case `CreateFranchise`)?

**Reply by weiznich** [reply]
> if i understand your answer correctly, you mean that in a derive-macro you have no way of knowing something like field _id(from the example) exists, because you only have access to the current struct (in this case CreateFranchise)?

That's correct

> yes, i ended up using .default_values(), but made this because it seemed weird to me (because initially i also wanted to have a similar structure across all tables and have it already available if the table should change)

It's likely possible to change the derive to generate something equivalent to the `default_values` function if the struct is empty. I would accept a PR for but it's not on my priority list, given that said function exists.
